xen/gnttbl: check the content of gnttbl entries when unmap happening

From: Oleksii Kurochko <oleksii.kurochko@gmail.com>

Signed-off-by: Oleksii Kurochko <oleksii.kurochko@gmail.com>

diff --git a/0001-xenconsoled-add-debug-log-in-possible-places-of-gran.patch b/0001-xenconsoled-add-debug-log-in-possible-places-of-gran.patch
new file mode 100644
index 0000000000..a7ee99466a
--- /dev/null
+++ b/0001-xenconsoled-add-debug-log-in-possible-places-of-gran.patch
@@ -0,0 +1,150 @@
+From 29981f8f83d1a75381f33a9e104020f9cce61f92 Mon Sep 17 00:00:00 2001
+From: Oleksii Kurochko <oleksii.kurochko@gmail.com>
+Date: Mon, 3 Nov 2025 10:12:36 +0100
+Subject: [PATCH] xenconsoled: add debug log in possible places of grant ref
+ unmapping
+
+Signed-off-by: Oleksii Kurochko <oleksii.kurochko@gmail.com>
+---
+ ...-limit-amount-of-p2m-lookup-messages.patch | 82 +++++++++++++++++++
+ tools/console/daemon/io.c                     | 17 ++++
+ 2 files changed, 99 insertions(+)
+ create mode 100644 0001-xen-arm-limit-amount-of-p2m-lookup-messages.patch
+
+diff --git a/0001-xen-arm-limit-amount-of-p2m-lookup-messages.patch b/0001-xen-arm-limit-amount-of-p2m-lookup-messages.patch
+new file mode 100644
+index 0000000000..371591e92d
+--- /dev/null
++++ b/0001-xen-arm-limit-amount-of-p2m-lookup-messages.patch
+@@ -0,0 +1,82 @@
++From ae9d097d874406a25f0d6d7cd64556f9f5e7104e Mon Sep 17 00:00:00 2001
++From: Oleksii Kurochko <oleksii.kurochko@gmail.com>
++Date: Fri, 31 Oct 2025 17:50:58 +0100
++Subject: [PATCH] xen/arm: limit amount of p2m lookup messages
++
++Print p2m entries information only when grant table mapping
++happens.
++
++Signed-off-by: Oleksii Kurochko <oleksii.kurochko@gmail.com>
++---
++ xen/arch/arm/mmu/p2m.c   | 19 ++++++++++++-------
++ xen/common/grant_table.c |  9 +++++++++
++ 2 files changed, 21 insertions(+), 7 deletions(-)
++
++diff --git a/xen/arch/arm/mmu/p2m.c b/xen/arch/arm/mmu/p2m.c
++index fafbafc7fc..09770e6b57 100644
++--- a/xen/arch/arm/mmu/p2m.c
+++++ b/xen/arch/arm/mmu/p2m.c
++@@ -432,6 +432,8 @@ static int p2m_next_level(struct p2m_domain *p2m, bool read_only,
++     return GUEST_TABLE_NORMAL_PAGE;
++ }
++ 
+++bool grant_table_req = false;
+++
++ /*
++  * Get the details of a given gfn.
++  *
++@@ -524,13 +526,16 @@ mfn_t p2m_get_entry(struct p2m_domain *p2m, gfn_t gfn,
++         if ( valid )
++             *valid = lpae_is_valid(entry);
++ 
++-        printk("%s: mfn(%#lx), v(%d), r(%d) w(%d) nx(%d)\n",
++-               __func__,
++-               mfn_x(mfn),
++-               lpae_is_valid(entry),
++-               entry.p2m.read,
++-               entry.p2m.write,
++-               entry.p2m.xn);
+++        if ( grant_table_req )
+++        {
+++            printk("%s: mfn(%#lx), v(%d), r(%d) w(%d) nx(%d)\n",
+++                __func__,
+++                mfn_x(mfn),
+++                lpae_is_valid(entry),
+++                entry.p2m.read,
+++                entry.p2m.write,
+++                entry.p2m.xn);
+++        }
++     }
++ 
++ out_unmap:
++diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
++index 22c7076189..cc6dad038b 100644
++--- a/xen/common/grant_table.c
+++++ b/xen/common/grant_table.c
++@@ -1367,6 +1367,8 @@ static void dump_grant_op(struct gnttab_map_grant_ref *op) {
++ 	dprintk(XENLOG_ERR, "dev_bus_addr: %016lx\n", op->dev_bus_addr);
++ }
++ 
+++extern bool grant_table_req;
+++
++ static long
++ gnttab_map_grant_ref(
++     XEN_GUEST_HANDLE_PARAM(gnttab_map_grant_ref_t) uop, unsigned int count)
++@@ -1386,7 +1388,14 @@ gnttab_map_grant_ref(
++ 
++ 	dprintk(XENLOG_ERR, "%s: before map_grant_ref()\n", __func__);
++ 	dump_grant_op(&op);
+++
+++
+++    grant_table_req = true;
+++
++         map_grant_ref(&op);
+++
+++    grant_table_req = false;
+++
++ 	dprintk(XENLOG_ERR, "%s: after map_grant_ref()\n", __func__);
++ 	dump_grant_op(&op);
++ 
++-- 
++2.51.0
++
+diff --git a/tools/console/daemon/io.c b/tools/console/daemon/io.c
+index f4e366ade0..0f8d68d5ea 100644
+--- a/tools/console/daemon/io.c
++++ b/tools/console/daemon/io.c
+@@ -680,7 +680,12 @@ static void console_unmap_interface(struct console *con)
+ 	if (con->interface == NULL)
+ 		return;
+ 	if (xgt_handle && con->ring_ref == -1)
++	{
++		dolog(LOG_ERR, ">>>>>>> %s:%s():L%d\n",
++		      __FILE__, __FUNCTION__, __LINE__);
++
+ 		xengnttab_unmap(xgt_handle, con->interface, 1);
++	}
+ 	else
+ 		xenforeignmemory_unmap(xfm_handle, con->interface, 1);
+ 	con->interface = NULL;
+@@ -719,9 +724,17 @@ static int console_create_ring(struct console *con)
+ 
+ 	/* If using ring_ref and it has changed, remap */
+ 	if (ring_ref != con->ring_ref && con->ring_ref != -1)
++	{
++		dolog(LOG_ERR, "%s:%s():L%d ring_ref has been changed? remapping?\n",
++		      __FILE__, __FUNCTION__, __LINE__);
++
+ 		console_unmap_interface(con);
++	}
+ 
+ 	if (!con->interface && xgt_handle && con->use_gnttab) {
++		dolog(LOG_ERR, "%s:%s():L%d ring_ref is console ring ref remapped?\n",
++		      __FILE__, __FUNCTION__, __LINE__);
++
+ 		/* Prefer using grant table */
+ 		con->interface = xengnttab_map_grant_ref(xgt_handle,
+ 			dom->domid, GNTTAB_RESERVED_CONSOLE,
+@@ -964,6 +977,10 @@ static void shutdown_domain(struct domain *d)
+ {
+ 	d->is_dead = true;
+ 	watch_domain(d, false);
++
++	dolog(LOG_ERR, ">>>>> %s:%s():L%d\n",
++		  __FILE__, __FUNCTION__, __LINE__);
++
+ 	console_iter_void_arg1(d, console_unmap_interface);
+ 	console_iter_void_arg1(d, console_close_evtchn);
+ }
+-- 
+2.51.1
+
diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
index cc6dad038b..e5a74e52b5 100644
--- a/xen/common/grant_table.c
+++ b/xen/common/grant_table.c
@@ -1680,11 +1680,15 @@ unmap_common_complete(struct gnttab_unmap_common *op)
     rcu_unlock_domain(rd);
 }
 
+
+static void gnttab_usage_print(struct domain *rd);
+
 static void
 unmap_grant_ref(
     struct gnttab_unmap_grant_ref *op,
     struct gnttab_unmap_common *common)
 {
+    struct domain *d;
     common->host_addr = op->host_addr;
     common->dev_bus_addr = op->dev_bus_addr;
     common->handle = op->handle;
@@ -1698,8 +1702,26 @@ unmap_grant_ref(
     printk("%s: host_addr(%#lx) dev_bus_addr(%#lx), handle(%#x)\n",
            __func__, op->host_addr, op->dev_bus_addr, op->handle);
 
+    rcu_read_lock(&domlist_read_lock);
+
+    printk("before:\n")
+
+    for_each_domain ( d )
+        gnttab_usage_print(d);
+
+    rcu_read_unlock(&domlist_read_lock);
+
     unmap_common(common);
     op->status = common->status;
+
+        rcu_read_lock(&domlist_read_lock);
+
+    printk("after:\n")
+
+    for_each_domain ( d )
+        gnttab_usage_print(d);
+
+    rcu_read_unlock(&domlist_read_lock);
 }
 
 
